Work items (#363)
OctopusDeploy/docs: docs/api-and-integration/metadata/index.md

---
title: Metadata and work-items
description: Configuring Octopus metadata and work-item integration.
position: 200
---

> This section details how to configure the custom metadata integration in Octopus. The metadata contains information about how a package was built and what it includes. The metadata includes work-items which represent the bugs and enhancements that are included in new releases of your software.

Reviewer: I made some minor tweaks to the suggestion.

![Package Details](package-detail.png)

If there is a specific issue tracker extension, e.g. Jira, enabled and configured that matches the work-items type the work-items will appear as a link to the issue tracking system. If not, the work-items will appear as plain text.

## Project Settings

The next step is to tell your Octopus project which packages in its deployment process contain metadata that you want included in the releases and deployments.

![Octopus Project work-item settings](octo-project.png)

In this example, the project is using the primary package from the step named Deploy. You can specify multiple packages if required. An example of where you might use this is a deployment process that includes steps for a Web App and a SQL database schema migration. The packages for both of those form the complete product and either or both could be contributing work-items.

## Releases and Deployments

When a release is created for the project, the selected package versions are used to collect the work-item list and it is shown with the release details.

![Release work-items](release-work-items.png)

> When this release is deployed the metadata will be contributed to the deployment. They will appear in the preview.

Author: ```suggestion
When this release is deployed the metadata from the release will be passed to the deployment. This metadata will be displayed in the deployment preview.
```

> ![Deployment preview](deploy-preview-work-items.png)

Author: More of the screen for context

> They will also appear on the task summary for the deployment.

Author: ```suggestion
They are also displayed on the task summary for the deployment.
```

> ![Deployment work-items](deploy-work-items.png)

Author: More screen for context please :)

In some scenarios this one to one between releases and deployments will be the norm. Teams using continuous deployment may see a rapid progression of versions through their environments and on to production.

> Many teams won't be operating like this though, many will accumulate a number of releases before promoting to the next environment. For example, they may merge pull requests for several fixes/features and then deploy the final release to test. This could even go through a couple of revolutions and then eventually a final release progresses from test to production. In this situation the deployment to production isn't just the work-items from that last release, it is the accumulation of all of the work-items in all of the releases since the last one that was deployed to production.

Author: ```suggestion
Many teams won't be operating like this though. It is common that a number of releases will be prepared and tested before promoting to the next environment. For example, they may merge pull requests for several fixes/features and then deploy the final release to test. This could even be repeated several times and then eventually a final release progresses from test to production. In this situation the deployment to production isn't just the work-items from that last release, it is _the accumulation of all of the work-items in all of the releases since the last one that was deployed to production_.
```

> This accumulation logic is how Octopus always determines the metadata list, it's always the accumulation since the last deployment to the given "scope". A scope in this context is the combination of deployment environment and tenant (if multi-tenancy is in play).

Author: ```suggestion
This accumulation logic is how Octopus always determines the metadata list, it's the accumulation since the last deployment to the given "scope". A scope in this context is the combination of deployment environment and tenant (if multi-tenancy is in play).
```

Can we use a different word that "scope"? That word is already used in other context, so may cause confusion.

Reviewer: I've been trying to think of an alternative. I agree scope gets used a lot. I can't think of anything but we could talk around it. The following perhaps?

This accumulation logic is how Octopus determines the metadata list, it's the accumulation since the last deployment to the *specified* environment and tenant (if multi-tenancy is in play).

> Given that you can initiate deployments to multiple "scopes" at once in Octopus it is quite conceivable, and expected, that you could see different work-items lists for each scope. Let's consider a couple of examples, first two tenants in the same environment. In this case if tenant A is on an earlier version than tenant B the the result list for tenant A would be the list for tenant B plus the additional work-items between the version it was on and the version tenant B was on.

Author: Not sure these examples are all that clear.

## Deploy a Release Step

The Octopus [deploy release step](https://g.octopushq.com/DeployReleaseStep) adds an interesting dimension to metadata accumulation. When you use this step Octopus treats each "child project" like it is a package. Following from this reasoning, it also treats the child project's release as a potential source for metadata.

On the project settings you can select a deploy release step and use it the same as any other package from the internal feed. When creating the releases in the "parent project" Octopus accumulates the release notes and metadata from the child projects just as it would for packages.

## Deployment Variables and the Email Step

During a deployment there are variables available for both the new release notes values and the work-items.

> The release notes variable is `Octopus.Deployment.Changes` and contains the release notes and work-items in JSON format. The structure is a JSON array of `ReleaseChange` objects matching the following C# class:

Author: Feels odd saying "its gonna be JSON, but here's the C# class definition".
Consider using a plain old JSON sample instead.

```csharp
public class ReleaseChanges
{
  public string Version { get; set; }
  public string ReleaseNotes { get; set; }
  public WorkItemLink[] WorkItems { get; set; }
}

public class WorkItemLink
{
    public string Id { get; set; }
    public string LinkUrl { get; set; }
    public string LinkText { get; set; }
}
```

There is an entry per release and it includes the release notes and the metadata for each of the packages in that release.

> The following example uses these variables to generate the HTML body for the Octopus email step.

Author: I'd give this its own heading

```html
Here are the notes:<br/>
#{each change in Octopus.Deployment.Changes}
<h2>#{change.Version}</h2>
#{change.ReleaseNotes}</br>
#{each workItem in change.WorkItems}
  #{if workItem.LinkUrl}
      <a href="#{workItem.LinkUrl}">#{workItem.LinkText}</a>
  #{/if}
  #{unless workItem.LinkUrl}
      #{workItem.LinkText}</br>
  #{/unless}
#{/each}
#{/each}
```

> Note that the if/unless is included for illustration purposes. The links can end up without a LinkUrl if the related extension (e.g. Jira) is disabled. If you know it's going to be enabled then just use the anchor.

Author: Consider removing - possibly too much information

## Issue Trackers

> If you are using an issue tracker like Jira, Octopus can also be integrated with it to provide information about the progress of work-items. For more details see

Author: ```suggestion
If you are using an issue tracker like Jira, Octopus can also be configured to provide information about the progress of work-items. For more details see
```

Actually, I'm not 100% sure what this sentence means. It sit pushing info to Jira? Or is it showing info from Jira in Octopus?

Many teams don't operate like this though. It is common for a number of releases to be prepared and tested before being promoting to the next environment. For example, they may merge pull requests for several fixes/features and then deploy the final release to test. This can be repeated several times and eventually a final release progresses from test to production. In this situation the deployment to production isn't just the work-items from that last release, it is _the accumulation of all of the work-items in all of the releases since the last one that was deployed to production_.

This accumulation logic is how Octopus always determines the metadata list, it's always the accumulation since the last deployment to the given "scope". A scope in this context is the combination of deployment environment and tenant (if [multi-tenancy](/docs/deployment-patterns/multi-tenant-deployments/index.md) is in play).

Given that you can initiate deployments to multiple "scopes" at once in Octopus it is expected that you could see different work-items lists for each scope. 

> For instance, if you have two tenants in the same environment, and **tenant A** is on an earlier version than **tenant B**, the resulting list for **tenant A** will be the list for **tenant B** plus the additional work-items between the version it was on and the version **tenant B** was on.

Reviewer: If I'm reading this correctly, I have a feeling we should make it clear this happens when we create a new release of tentant A and deploy it so that it is now on the same version as tenant B. The suggestion below attempts to make that clear.

```suggestion
For instance, if you have two tenants in the same environment, and **tenant A** is on an earlier version than **tenant B**, when **tenant A** moves to the same version as **tenant B**, the resulting list of work-items for **tenant A** will be the list for **tenant B** plus the additional work-items between the version **tenant A** was on and the version **tenant B** was on.
```
